File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        self.maxSum = 0

        def dfs(node):
            if not node:
                return (True, float('inf'), float('-inf'), 0)

            leftIsBST, leftMin, leftMax, leftSum = dfs(node.left)
            rightIsBST, rightMin, rightMax, rightSum = dfs(node.right)

            if leftIsBST and rightIsBST and leftMax < node.val < rightMin:
                currentSum = node.val + leftSum + rightSum
                self.maxSum = max(self.maxSum, currentSum)
                return (True, min(leftMin, node.val), max(rightMax, node.val), currentSum)
            else:
                return (False, 0, 0, 0)

        dfs(root)
        return self.maxSum

        

File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum = 0;

    tuple<bool, int, int, int> dfs(TreeNode* node) {
        if (!node)
            return {true, INT_MAX, INT_MIN, 0};

        auto [leftBST, leftMin, leftMax, leftSum] = dfs(node->left);
        auto [rightBST, rightMin, rightMax, rightSum] = dfs(node->right);

        if (leftBST && rightBST && leftMax < node->val && node->val < rightMin) {
            int currSum = node->val + leftSum + rightSum;
            maxSum = max(maxSum, currSum);
            return {true, min(leftMin, node->val), max(rightMax, node->val), currSum};
        } else {
            return {false, 0, 0, 0};
        }
    }

    int maxSumBST(TreeNode* root) {
        dfs(root);
        return maxSum;
    }
};
